### ISSUE TITLE
net is not-defined when noCuda

### ISSUE THREAD

**Issue by csypeng** [ChenRocks/fast_abs_rl#11]
Line156 of train_abstractor.py:
if args.cuda:
        net = net.cuda()
pipeline = BasicPipeline(meta['net'], net,
                   train_batcher, val_batcher, args.batch, val_fn,
                   criterion, optimizer, grad_fn)

Since there is no cuda, net is not-defined. Therefore, the train_step function of BasicPipeline in training.py will produce a Segmentation fault.

**Comment by ChenRocks** [comment]
If your machine does not support cuda GPU than `args.cuda` should automatically be False. Hence the line `net = net.cuda()` will not be executed. Could you give more detail on the segfault?

**Comment by csypeng** [comment]
I dig more and find the segment fault is not because of Line156 of train_abstractor.py.

I find torch.matmul(enc_art, self._attn_wm) at Line94 of summ.py through a segfault. I find the shape of enc_art is 27, 32, 512 and the shape of self._attn_wm is 512, 256 (when I using the default parameters of train_abstractor.py). So, I want to see if it is the problem of too large tensors, as I just have CPU. I open a new terminal, enter the python command line environment and do matmul of a tensor of 27, 32, 512 and a tensor of 512, 256. I find that still segfault. Therefore, I guess it is because tensor is too large. 

Therefore, I set the dim of word vector to be small =12 (simply delete the last digit of 8 of the default 128) and the --n_hidden=25. Line94 of summ.py can pass. But, loss.backward() at Line104 of training.py through a segfault. :(

**Comment by ChenRocks** [comment]
If you get a *segmentation fault* in python, I guess it's due to trying to allocate memory beyond system/hardware limit. Interestingly, even with a cpu-only macbook pro (16G RAM) I was able to perform `matmul` on tensors of size [27, 32, 512] and [512, 256]. Could you provide the information of the system you use?

However, my best suggestion is to use some modern server with GPU if you want to get some good quality of experiments with this repository. An AWS p2.xlarge should be good to go and not too costly.

**Comment by csypeng** [comment]
If your cpu-only macbook pro (16G RAM) can perform matmul on tensors of size [27, 32, 512] and [512, 256], I think some library is not well installed at my side, because that my cpu-server has 300G memory. 

I will try the AWS p2.xlarge as well.

**Comment by ChenRocks** [comment]
I would suspect there is some system limit on how much memory a single process can use in your server. I will use the `ulimit -a` command to find this out first.

**Comment by csypeng** [comment]
core file size          (blocks, -c) 0
data seg size           (kbytes, -d) unlimited
scheduling priority             (-e) 0
file size               (blocks, -f) unlimited
pending signals                 (-i) 1545511
max locked memory       (kbytes, -l) 64
max memory size         (kbytes, -m) unlimited
open files                      (-n) 40000
pipe size            (512 bytes, -p) 8
POSIX message queues     (bytes, -q) 819200
real-time priority              (-r) 0
stack size              (kbytes, -s) 8192
cpu time               (seconds, -t) unlimited
max user processes              (-u) 1545511
virtual memory          (kbytes, -v) unlimited
file locks                      (-x) unlimited

Is it okay? 
max locked memory       (kbytes, -l) 64   
max memory size         (kbytes, -m) unlimited

**Comment by ChenRocks** [comment]
Well, the output looks normal. Then it might be some package installation issue as you previously suggested. FYI, I use anaconda python for the package dependencies. Maybe you can try to create a virtual env and reinstall everything. 
```bash
conda create -n fast-abs python=3.6
source activate fast-abs
# install packages
```

**Comment by ChenRocks** [comment]
closing due to no further discussion.
